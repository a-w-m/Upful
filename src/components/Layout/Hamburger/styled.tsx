import styled from "styled-components"
import device from "../mediaQuery"

export const Container = styled.div<{ open: boolean }>`
  display: flex;
  width: ${({ open }) => (open ? "100%" : "auto")};
  justify-content: ${({ open }) => (open ? "flex-end" : "flex-start")};
`

export const Button = styled.button<{ open: boolean }>`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  width: 3rem;
  height: 2rem;
  background: transparent;
  border: none;
  cursor: pointer;
  padding: 0;
  z-index: 10;
  padding-bottom: 0.1rem;

  &:focus {
    outline: none;
  }

  span {
    width: 2.4rem;
    height: 0.2rem;
    background: #000;
    border-radius: 1rem;
    transition: all 0.3s linear;
    position: relative;
    transform-origin: 0.3rem;

    :first-child {
      transform: ${({ open }) => (open ? "rotate(45deg)" : "rotate(0)")};
    }

    :nth-child(2) {
      opacity: ${({ open }) => (open ? "0" : "1")};
      transform: ${({ open }) => (open ? "translateX(2rem)" : "translateX(0)")};
    }

    :nth-child(3) {
      transform: ${({ open }) => (open ? "rotate(-45deg)" : "rotate(0)")};
    }
  }

  @media ${device.tablet} {
    display: none;
  }
`
