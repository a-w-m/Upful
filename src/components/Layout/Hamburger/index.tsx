import React, { SetStateAction, Dispatch } from "react"
import { Container, Button } from "./styled"

interface HamburgerProps {
  open: boolean
  setOpen: Dispatch<SetStateAction<boolean>>
}

const Hamburger: React.FC<HamburgerProps> = props => {
  const { open, setOpen } = props

  return (
    <Container open={open} onClick={() => setOpen(prev => !prev)}>
      <Button
        type="button"
        aria-label="open navigation"
        aria-controls="link-list"
        aria-expanded="false"
        open={open}
      >
        <div />
        <div />
        <div />
      </Button>
    </Container>
  )
}

export default Hamburger
