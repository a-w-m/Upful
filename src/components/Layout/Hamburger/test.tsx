import React from "react"
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import Hamburger from "."

describe("Hamburger component", () => {
  const setOpen = jest.fn()

  it("should match snapshot", () => {
    render(<Hamburger open={false} setOpen={setOpen} />)
    const button = screen.getByRole("button")
    expect(button).toMatchSnapshot()
  })

  it("should handle click", async () => {
    render(<Hamburger open={false} setOpen={setOpen} />)
    const user = userEvent.setup()
    const button = screen.getByRole("button")
    await user.click(button)
    expect(setOpen).toHaveBeenCalledTimes(1)
  })

  it("should apply style when open", () => {
    render(<Hamburger open={true} setOpen={setOpen} />)
    const menu = document.querySelectorAll("span")
    expect(menu[0]).toHaveStyle({ transform: "rotate(45deg)" })
    expect(menu[1]).toHaveStyle({ opacity: "0", transform: "translateX(2rem)" })
    expect(menu[2]).toHaveStyle({ transform: "rotate(-45deg)" })
  })

  it("should apply style when not open", () => {
    render(<Hamburger open={false} setOpen={setOpen} />)
    const menu = document.querySelectorAll("span")
    expect(menu[0]).toHaveStyle({ transform: "rotate(0)" })
    expect(menu[1]).toHaveStyle({ opacity: "1", transform: "translateX(0)" })
    expect(menu[2]).toHaveStyle({ transform: "rotate(0)" })
  })
})
