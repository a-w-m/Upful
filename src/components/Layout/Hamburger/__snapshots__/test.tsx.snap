// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger component component tree matches snapshot 1`] = `
<button
  aria-controls="link-list"
  aria-expanded="false"
  aria-label="open navigation"
  class="sc-bdfBwQ jrcbxy"
  open=""
  type="button"
>
  <div />
  <div />
  <div />
</button>
`;
